modules/common/secret/secret.go

```diff
@@ -385,3 +386,70 @@ func GetDataFromSecret(
 
 	return data, ctrl.Result{}, nil
 }
+
+// EnsureSecretData - ensures that the Secret object exists and the expected fields
+// are in the Secret. It returns a hash of the values of the expected fields.
+func EnsureSecretData(
+	ctx context.Context,
+	secretName types.NamespacedName,
+	expectedFields []string,
+	reader client.Reader,
+	Updater condition.Updater,
+	requeueTimeout time.Duration,
+) (string, ctrl.Result, error) {
+	secret := &corev1.Secret{}
+	err := reader.Get(ctx, secretName, secret)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// TODO(gibi): Change the message to state which input
+			// (i.e. Secret with a given name) is missing
+			Updater.Set(condition.FalseCondition(
+				condition.InputReadyCondition,
+				condition.RequestedReason,
+				condition.SeverityInfo,
+				condition.InputReadyWaitingMessage))
+			return "",
+				ctrl.Result{RequeueAfter: requeueTimeout},
+				fmt.Errorf("OpenStack secret %s not found", secretName)
+		}
+		Updater.Set(condition.FalseCondition(
+			condition.InputReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.InputReadyErrorMessage,
+			err.Error()))
+		return "", ctrl.Result{}, err
+	}
+
+	// collect the secret values the caller expects to exist
+	values := [][]byte{}
+	for _, field := range expectedFields {
+		val, ok := secret.Data[field]
+		if !ok {
+			err := fmt.Errorf("field %s not found in Secret %s", field, secretName)
+			Updater.Set(condition.FalseCondition(
+				condition.InputReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				condition.InputReadyErrorMessage,
+				err.Error()))
+			return "", ctrl.Result{}, err
+		}
+		values = append(values, val)
+	}
+
+	// TODO(gibi): Do we need to watch the Secret for changes?
```

Comment:
While it was still in nova only I was OK to keep this TODO in place, but as soon as we move this to lib-common this becomes complicated. The goal is valid to eventually watch the Secret for changes so that we can pick up when the missing field is eventually provided. But we cannot easily add such watch via this lib-common function. So the TODO will not be resolved here but probably in every individual operator.

So at least I would change this TODO here to express this.

Reply:
so in nova yes we do but this is out os the scope fo this function

we should have a separate function to do this if needed.

the comment should be removed i think.

Reply:
But if we remove the TODO comment here then we loose the TODO without resolving it. This is why I'm torn on this.

Reply:
I think we should have a convention base on labels for CM/Secret objects to watch which one does not own. Is this what is meant by this comment? The creator would have to set the label/labels, could be something like `global-watcher` which all operators watch, `glance-watcher`, `nova-watcher` for service specific

Reply:
we can keep the todo in the nova code when we adapt to using it form lib common.

@stuggi 
well we could do that or we can just watch the secrets/configmaps that are provided as inputs to the service crs.

it would be nicer if we could subscribe to the one we cared about from the input set rather then requiring the OpenStack operator to set those labels.

Reply:
I don't know exactly how to do it in go, but maybe we can turn this into an abstract class where people would need to implement the method for watching the secret

Reply:
found this: https://golangbyexample.com/go-abstract-class/

Reply:
> I think we should have a convention base on labels for CM/Secret objects to watch which one does not own. Is this what is meant by this comment? The creator would have to set the label/labels, could be something like `global-watcher` which all operators watch, `glance-watcher`, `nova-watcher` for service specific

I haven't had time to dig into solving the TODO itself so I don't know what would be a good solution. I assume that as the given operator gets the name of the Secret via the CRD field the easiest way would be to watch the Secret by name. But I don't know if that is possible or practical.

Reply:
@stuggi thanks, that shows some downside of our CRD designs. So if it is hard to watch secrets by name as the name is dynamic and the watch is easiest to set statically in `SetupWithManager` then I think we should reconsider passing Secret by name to our CRDs and come up with something better suited for watching for changes.

Reply:
ah you can also do it with a spec field like here https://book.kubebuilder.io/reference/watching-resources/externally-managed.html

Reply:
@stuggi cool then we can keep the Secret name field on our CRD and set up a watcher still.

Back to the original topic. I think I'm OK with removing the TODO from here and adding a sentence to the doc of this function stating that it is the responsibility of the caller to make sure it sets up a watcher for the Secret to get informed about any changes.